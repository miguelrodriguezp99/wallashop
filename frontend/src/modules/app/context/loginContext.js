import { createContext, useState } from "react";

// Este es el que tenemos que consumir
export const LogginContext = createContext();

// Este es el que nos provee de acceso al contexto
export function LogginProvider({ children }) {
	const [token, setToken] = useState(null);
	const [image, setImage] = useState(null);
	const [user, setUser] = useState({});

	return (
		<LogginContext.Provider
			value={{
				token,
				setToken,
				image,
				setImage,
				user,
				setUser,
			}}
		>
			{children}
		</LogginContext.Provider>
	);
}
