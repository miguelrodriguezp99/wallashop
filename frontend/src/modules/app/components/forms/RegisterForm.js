import { Button } from "@nextui-org/react";
import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "sonner";
import { signUp } from "../../../../backend/userService.js";
import { config } from "../../../../config/constants.js";
import { LogginContext } from "../../context/loginContext.js";
import { uploadFile } from "../../firebase/config.js";
import Input from "../inputs/LoginInput.js";

const RegisterForm = () => {
	const avatarUrl = "https://random.imagecdn.app/500/500";
	const navigate = useNavigate();
	const [userName, setUserName] = useState("");
	const [email, setEmail] = useState("");
	const [password, setPassword] = useState("");
	const [firstName, setFirstName] = useState("");
	const [lastName, setLastname] = useState("");
	const [isVisible, setIsVisible] = React.useState(false);
	const [avatar, setAvatar] = useState(avatarUrl);
	const { setImage } = useContext(LogginContext);
	const toggleVisibility = () => setIsVisible(!isVisible);

	function getParams() {
		const formData = {
			userName,
			password,
			firstName,
			lastName,
			email,
			avatar,
		};
		return formData;
	}

	const onSuccess = () => {
		setImage(avatar);
		console.log("he llegado al onSuccess");
		toast.success("User created successfully");
		navigate("/wallashop/login");
	};

	const onErrors = (error) => {
		toast.error("Error al crear el usuario.");
		console.log(error);
	};

	const reauthenticateUser = () => {
		console.log(
			"Se requiere reautenticación. Mostrar cuadro de diálogo de inicio de sesión.",
		);
	};

	const handleSubmit = () => {
		const user = getParams();
		signUp(user, onSuccess, onErrors, reauthenticateUser);
	};

	const handleSubmitAvatar = async (files) => {
		const avatar = files[0];
		console.log("avatar: " + avatar);
		const result = await uploadFile(avatar);
		const fullPath = result.metadata.fullPath;
		const route = `https://firebasestorage.googleapis.com/v0/b/${config.FIREBASE_PROYECT}.appspot.com/o/${fullPath}?alt=media`;
		localStorage.setItem("avatar", route);
		console.log("route: " + route);
		setAvatar(route);
	};

	return (
		<>
			<div className="flex flex-col items-center justify-center min-h-screen bg-black">
				<div className="border border-solid border-gray-200 rounded-2xl text-center text-gray-200 font-helvetica py-10 px-16 w-500 mt-0 hover:border-blue-300">
					<h1 className="text-7xl font-bold bg-gradient-to-r from-blue-500 to-blue-700 text-transparent bg-clip-text">
						SIGN UP
					</h1>
					<form onSubmit={() => handleSubmit()}>
						<label htmlFor="username">
							<Input
								id="username"
								type="text"
								name="username"
								value={userName}
								placeholder={"Username"}
								fn={setUserName}
							/>
						</label>
						<label htmlFor="email">
							<Input
								id="email"
								type="text"
								name="email"
								value={email}
								placeholder={"Email"}
								fn={setEmail}
							/>
						</label>
						<label htmlFor="password">
							<Input
								id = "password"
								label="Password"
								name="password"
								value={password}
								placeholder={"Password"}
								fn={setPassword}
								endContent={
									<button
										className="focus:outline-none"
										type="button"
										onClick={toggleVisibility}
									></button>
								}
								type={isVisible ? "text" : "password"}
								className="max-w-xs"
							/>
						</label>
						<label htmlFor="firstname">
							<Input
								id="firstname"
								type="text"
								name="firstName"
								value={firstName}
								placeholder={"FirstName"}
								fn={setFirstName}
							/>
						</label>
						<label htmlFor="lastname">
							<Input
								id="lastname"
								type="text"
								name="lastname"
								value={lastName}
								placeholder={"LastName"}
								fn={setLastname}
							/>
						</label>
						<label
							htmlFor="formFile"
							className="mt-0 inline-block dark:text-neutral-200 text-white"
						>
							Avatar
						</label>
						<input
							className="relative m-0 block w-full min-w-0 flex-auto rounded border border-solid border-neutral-300 bg-clip-padding px-3 py-[0.32rem] text-base font-normal text-neutral-700 transition duration-300 ease-in-out file:-mx-3 file:-my-[0.32rem] file:overflow-hidden file:rounded-none file:border-0 file:border-solid file:border-inherit file:bg-neutral-100 file:px-3 file:py-[0.32rem] file:text-neutral-700 file:transition file:duration-150 file:ease-in-out file:[border-inline-end-width:1px] file:[margin-inline-end:0.75rem] hover:file:bg-neutral-200 focus:border-primary focus:text-neutral-700 focus:shadow-te-primary focus:outline-none dark:border-neutral-600 dark:text-neutral-200 dark:file:bg-neutral-700 dark:file:text-neutral-100 dark:focus:border-primary mt-5"
							type="file"
							id="formFile"
							onChange={(e) => {
								handleSubmitAvatar(e.target.files);
							}}
						/>

						<div className="form-buttons mt-5">
							<Button
								className="text-white ml-2 mr-2"
								onPress={() => handleSubmit()}
								as={Button}
								color="primary"
								href="#"
								variant="ghost"
							>
								Sign Up
							</Button>

							<Button
								className="text-white ml-2 mr-2"
								onPress={() => navigate("/wallashop/home")}
								as={Button}
								color="primary"
								href="#"
								variant="ghost"
							>
								Home
							</Button>
						</div>
					</form>
				</div>
			</div>
		</>
	);
};

export default RegisterForm;
