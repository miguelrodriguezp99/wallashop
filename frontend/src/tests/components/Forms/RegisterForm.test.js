import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import RegisterForm from '../../../modules/app/components/forms/RegisterForm';
import { LogginContext } from '../../../modules/app/context/loginContext'; 
import { useLocation } from 'react-router-dom';
import '@testing-library/jest-dom';

jest.mock('../../../modules/app/firebase/config.js', () => ({
}));

describe('RegisterForm component', () => {
    it('renders all components correctly', () => {
        const fakeLogginContext = {
            login: jest.fn(),
            setToken: jest.fn(),
            setUser: jest.fn(),
            setImage: jest.fn(),
        };

        const TestComponent = () => {
            testLocation = useLocation();
            return null;
        };
            
        let testHistory, testLocation;
        render(
            <MemoryRouter initialEntries={['/register']}>
            <LogginContext.Provider value={fakeLogginContext}>
                <Routes>
                <Route path="/register" element={<RegisterForm />} />
                <Route path="*" element={<TestComponent />} />  
                </Routes>
            </LogginContext.Provider>
            </MemoryRouter>
        );

        expect(screen.getByPlaceholderText('Username')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Email')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();

        const signUpButton = screen.getByText('Sign Up');
        const homeButton = screen.getByText('Home');

        expect(signUpButton).toBeInTheDocument();
        expect(homeButton).toBeInTheDocument();

        fireEvent.click(homeButton);
        expect(testLocation.pathname).toBe('/wallashop/home');
    });

    it('matches snapshot', () => {
        const fakeLogginContext = {
            login: jest.fn(),
            setToken: jest.fn(),
            setUser: jest.fn(),
            setImage: jest.fn(),
        };

        const { asFragment } = render(
            <MemoryRouter>
            <LogginContext.Provider value={fakeLogginContext}>
                <RegisterForm />
            </LogginContext.Provider>
            </MemoryRouter>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});