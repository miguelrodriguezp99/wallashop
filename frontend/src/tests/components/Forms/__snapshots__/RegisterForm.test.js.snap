// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center justify-center min-h-screen bg-black"
  >
    <div
      class="border border-solid border-gray-200 rounded-2xl text-center text-gray-200 font-helvetica py-10 px-16 w-500 mt-0 hover:border-blue-300"
    >
      <h1
        class="text-7xl font-bold bg-gradient-to-r from-blue-500 to-blue-700 text-transparent bg-clip-text"
      >
        SIGN UP
      </h1>
      <form>
        <label
          for="username"
        >
          <div
            class="form__group field"
          >
            <input
              class="form__field"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
            <label
              class="form__label"
              for="name"
            >
              Username
            </label>
          </div>
        </label>
        <label
          for="email"
        >
          <div
            class="form__group field"
          >
            <input
              class="form__field"
              name="email"
              placeholder="Email"
              type="text"
              value=""
            />
            <label
              class="form__label"
              for="name"
            >
              Email
            </label>
          </div>
        </label>
        <label
          for="password"
        >
          <div
            class="form__group field"
          >
            <input
              class="form__field"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <label
              class="form__label"
              for="name"
            >
              Password
            </label>
          </div>
        </label>
        <label
          for="firstname"
        >
          <div
            class="form__group field"
          >
            <input
              class="form__field"
              name="firstName"
              placeholder="FirstName"
              type="text"
              value=""
            />
            <label
              class="form__label"
              for="name"
            >
              FirstName
            </label>
          </div>
        </label>
        <label
          for="lastname"
        >
          <div
            class="form__group field"
          >
            <input
              class="form__field"
              name="lastname"
              placeholder="LastName"
              type="text"
              value=""
            />
            <label
              class="form__label"
              for="name"
            >
              LastName
            </label>
          </div>
        </label>
        <label
          class="mt-0 inline-block dark:text-neutral-200 text-white"
          for="formFile"
        >
          Avatar
        </label>
        <input
          class="relative m-0 block w-full min-w-0 flex-auto rounded border border-solid border-neutral-300 bg-clip-padding px-3 py-[0.32rem] text-base font-normal text-neutral-700 transition duration-300 ease-in-out file:-mx-3 file:-my-[0.32rem] file:overflow-hidden file:rounded-none file:border-0 file:border-solid file:border-inherit file:bg-neutral-100 file:px-3 file:py-[0.32rem] file:text-neutral-700 file:transition file:duration-150 file:ease-in-out file:[border-inline-end-width:1px] file:[margin-inline-end:0.75rem] hover:file:bg-neutral-200 focus:border-primary focus:text-neutral-700 focus:shadow-te-primary focus:outline-none dark:border-neutral-600 dark:text-neutral-200 dark:file:bg-neutral-700 dark:file:text-neutral-100 dark:focus:border-primary mt-5"
          id="formFile"
          type="file"
        />
        <div
          class="form-buttons mt-5"
        >
          <button
            class="group tap-highlight-transparent transition-transform-colors-opacity data-[hover=true]:opacity-hover z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium bg-transparent px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 rounded-medium [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none border-primary hover:!text-primary-foreground hover:!bg-primary text-white ml-2 mr-2"
            href="#"
            role="button"
            tabindex="0"
            type="button"
          >
            Sign Up
          </button>
          <button
            class="group tap-highlight-transparent transition-transform-colors-opacity data-[hover=true]:opacity-hover z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium bg-transparent px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 rounded-medium [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none border-primary hover:!text-primary-foreground hover:!bg-primary text-white ml-2 mr-2"
            href="#"
            role="button"
            tabindex="0"
            type="button"
          >
            Home
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
